### PR TITLE
Storage Improvements

### PR DESCRIPTION
**Pull request by tsandall**

### CODE REVIEW

File: storage/inmem/txn.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+			return nil
+		}
+	}
+
+	path, remove, value, err := validate(txn.db.data, op, path, 0, value)
+	if err != nil {
+		return err
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: using https://golang.org/pkg/container/list/ would clarify the code quite a bit.

File: storage/inmem/txn.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	value  interface{}
+	next   *snapshot
+	prev   *snapshot
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) assert(op storage.PatchOp, path storage.Path, value interface{}) error {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: the function name is rather confusing.

File: storage/inmem/txn.go [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+	"fmt"
+	"strconv"
+
+	"github.com/open-policy-agent/opa/storage"
+)
+
+type transaction struct {
+	xid      uint64
+	write    bool
+	log      *snapshot
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: please describe the contents of log: it is a sequence of snapshots, in what order?

File: storage/inmem/txn.go [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+	db       *store
+	policies map[string][]byte
+}
+
+type snapshot struct {
+	path   storage.Path
+	remove bool
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: an overall comment for snapshot would be good: what is the high-level purpose of each field.

File: storage/inmem/txn.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+	prev   *snapshot
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) assert(op storage.PatchOp, path storage.Path, value interface{}) error {
+
+	if !txn.write {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: I did not realize right away you don't support upgrading read transactions to read/write transactions. Worth of mentioning in the top transaction comment.

File: storage/inmem/txn.go [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+			Code:    storage.InvalidTransactionErr,
+			Message: "data write during read transaction",
+		}
+	}
+
+	if len(path) == 0 {
+		return txn.assertRoot(op, value)
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: High level comment here: start from the most recent snapshot, go backwards until a prefix match.

File: storage/inmem/txn.go [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+		if curr.path.Equal(path) {
+			if curr.remove {
+				if op != storage.AddOp {
+					return notFoundError(path)
+				}
+			}
+			curr.stale = true
+			break
+		} else if curr.path.HasPrefix(path) {
+			curr.stale = true
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: This stale case definitely needs a comment: a top level comment around "stale bool" would have been enough.

Comment (koponen-styra) [inline storage/inmem/txn.go]: Perhaps stale -> masked?

File: storage/inmem/txn.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+		} else if path.HasPrefix(curr.path) {
+			if curr.remove {
+				return notFoundError(path)
+			}
+			suffix := path[len(curr.path):]
+			path, remove, value, err := validate(curr.value, op, suffix, 0, value)
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: I'm missing more descriptive function name here: validate tells me nothing about the operation executed.

File: storage/inmem/txn.go (hunk 6 above)
Comment (koponen-styra) [inline storage/inmem/txn.go]: Also the notion of modifying existing snapshot is not typical: with your typical transaction log, you don't modify the "ops" after their insertion to log. Worth of mentioning somewhere.

File: storage/inmem/txn.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+	if txn.log != nil {
+		txn.log.prev = curr
+	}
+
+	txn.log = curr
+	return nil
+}
+
+func (txn *transaction) assertRoot(op storage.PatchOp, value interface{}) error {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: assertRoot is equally confusing.

File: storage/inmem/txn.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+	if !txn.write {
+		return &storage.Error{
+			Code:    storage.InvalidTransactionErr,
+			Message: "data write during read transaction",
+		}
+	}
+
+	if len(path) == 0 {
+		return txn.assertRoot(op, value)
+	}
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: more of a nit but I don't see a reason why you special case the handling of root patching.

File: storage/inmem/txn.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if len(curr.path) > 0 {
+			apply(txn.db.data, curr.path, curr.remove, curr.value)
+		} else {
+			txn.db.data = curr.value.(map[string]interface{})
+		}
+		result.SetDataChanged()
+	}
+	for id, bs := range txn.policies {
+		if bs == nil {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: this use of nil as a sign of remove should be mentioned in snapshot struct comment.

File: storage/inmem/txn.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+		}
+		if curr.path.HasPrefix(path) {
+			merge = append(merge, curr)
+		}
+	}
+
+	data, err := ptr(txn.db.data, path)
+
+	if err != nil {
+		if !storage.IsNotFound(err) || len(merge) == 0 {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: Hmm, this if looks peculiar: I would assumed && here. Pointer path might exist only after patches have been applied, or?

File: storage/inmem/txn.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+		if bs, ok := txn.policies[id]; !ok || bs != nil {
+			ids = append(ids, id)
+		}
+	}
+	for id, bs := range txn.policies {
+		if bs != nil {
+			if _, ok := txn.db.policies[id]; !ok {
+				ids = append(ids, id)
+			}
+		}
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: shouldn't this remove the id from the ids list  if bs == nil? the ids list could have duplicates, is that OK?

Comment (koponen-styra) [inline storage/inmem/txn.go]: if this is to be used as a "super set" of the actual list, i.e. the caller is expected to verify the existence of the policy afterwards with getPolicy, the function name/comment should reflect that.

File: storage/inmem/txn.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+		return nil, notFoundErrorf("policy id %q", id)
+	}
+	if exist, ok := txn.db.policies[id]; ok {
+		return exist, nil
+	}
+	return nil, notFoundErrorf("policy id %q", id)
+}
+
+func (txn *transaction) setPolicy(id string, bs []byte) error {
+	if !txn.write {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: pondering: these universal if txn.write checks in the beginning of functions make me wonder if you should just have two transaction impls, read and write.

Comment (koponen-styra) [inline storage/inmem/txn.go]: it would make the non-upgradeablity more obvious too.

File: storage/inmem/txn.go [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+		for k := range val {
+			cpy[k] = deepCopy(val[k])
+		}
+		return cpy
+	default:
+		return val
+	}
+}
+
+func getUnsafe(data interface{}, path storage.Path) interface{} {
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: not sure if the unsafe here captures the right intent. at least I was assuming golang unsafe pkg right away...

File: storage/inmem/txn.go [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+		return validateObject(data, op, path, idx, value)
+
+	case []interface{}:
+		return validateArray(data, op, path, idx, value)
+
+	case nil, bool, json.Number, string:
+		return nil, false, nil, notFoundError(path)
+	}
+
+	return nil, false, nil, fmt.Errorf("invalid data type")
```
Comment (koponen-styra) [inline storage/inmem/txn.go]: is there a error type for invalid data like there is for not found?

File: storage/inmem/txn.go (hunk 12 above)
Comment (tsandall) [inline storage/inmem/txn.go]: I think it's correct as is; the logic is a bit confusing though.

If && was used then non-NotFound errors would only be returned if there was a matching snapshot for path.

The current logic (in pseudocode) is:

- If non-NotFound error, return err
- If NotFound error AND no matching snapshots, return err

File: storage/inmem/txn.go (hunk 13 above)
Comment (tsandall) [inline storage/inmem/txn.go]: Ugh, again I think it's correct as is, but the logic is written in a confusing manner...

The code never adds an id to the list if bs == nil.

First loop adds ids from db that are either not in the txn (`!ok`) or in the txn but not removed (`bs != nil`).

Second loop adds ids from txn that are not removed (`bs != nil`) and not in db (`!ok`).

Whew. Can definitely clean this up.

File: storage/inmem/txn.go (hunk 12 above)
Comment (tsandall) [inline storage/inmem/txn.go]: Reviewing this logic w/ severe jet lag was not a good idea. This part is indeed buggy. Will fix.

**Pull request merged**